Turotial
------------

Test your installation by invoking pytest, in a terminal in the root directory of HBVouroboros:

.. code-block:: console

   $ pytest

This will run the pipeline with default paramters and confirm that the required output files are produced andmutations are correctly called.


RNAsim is a bare-bone RNA read simulator used prinicipally to generate reads for automatic integrations runs of the pipeline via github actions.
The user can run the program using reads generated by RNAsim by specifying the relevant parameters in the ``config/config.yaml``


.. code-block:: python

   ## use either way to indicate input files: either sample_annotation, or biokit_outdir
   ## sample_annotation should be a tab-delimited file containing at least four columns (#ID, GROUP, FASTQ1, FASTQ2)
   sample_annotation: RNAsim2/output/sampleAnnotation
   biokit_outdir: None

   ## options for read trimming with Trimmomatic
   trim_reads: False
   illumina_clip_file: 'resources/trim_reads/primers.fasta'
   illumina_clip_opts: ':2:30:10:2:keepBothReads'
   # trimmomatic steps (excluding ILLUMINACLIP)
   trimmomatic_steps: 'LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:35 CROP:140'
   
   ## Additional modes
   doInputRef: False
   inputRef: 'gnl|hbvnuc|AB064313_FT00000_C-G'
   doPerSamp: False
   
   ## RNAsim
   doSim: False
   sample_annotation_sm: RNAsim2/output/sampleAnnotation
   genomeId: gnl|hbvnuc|GQ924620_FT00000_C-C
   sampNum: "165000"
   pairedEndDist: "250"
   readLen: "150"
   mt: -mt
   mp: -mp 
   mtPos: "100"

The input reads are listed in a ``sampleAnnotation`` file. The corresponding parameter in the default config file point to the output directory of RNAsim, you can change this to point to your input. Below you see the default ``sampleAnnotation``, which has the minimum required number of enteries, 4, for each sample

.. code-block:: python

   #ID    GROUP FASTQ1  FASTQ2
   simSample1   control RNAsim2/simSamples/simSample-1_1.fastq.gz       RNAsim2/simSamples/simSample-1_2.fastq.gz
   simSample2   control RNAsim2/simSamples/simSample-2_1.fastq.gz       RNAsim2/simSamples/simSample-2_2.fastq.gz

The program can be run in three different modes: 1) infref: the reference genome is inferred from the aggregation of all samples, 2) inpt: samples are compared against a given genotype of HBV, and 3) per_Samp: the reference genome is inferred for each sample. By default all three modes are active. 
 


Once the input files are specified and the run-mode is selected, run the pipeline by typing the following in a terminla at the root directory of HBVouroboros

.. code-block:: console

   $ snakemake --cores 1 --use-conda

This will execute the pipeline localy. Please refer to snakemake documentation for job submission via SLURM or other schedulers.
