Turotial
------------

HBVouroboros is a snakemake pipeline and its parameters are listed in ``config/config.yaml``, shown below with default paramters.



.. code-block:: python

   ## use either way to indicate input files: either sample_annotation, or biokit_outdir
   ## sample_annotation should be a tab-delimited file containing at least four columns (#ID, GROUP, FASTQ1, FASTQ2)
   sample_annotation: RNAsim2/output/sampleAnnotation
   biokit_outdir: None

   ## options for read trimming with Trimmomatic
   trim_reads: False
   illumina_clip_file: 'resources/trim_reads/primers.fasta'
   illumina_clip_opts: ':2:30:10:2:keepBothReads'
   # trimmomatic steps (excluding ILLUMINACLIP)
   trimmomatic_steps: 'LEADING:20 TRAILING:20 SLIDINGWINDOW:4:15 MINLEN:35 CROP:140'
   
   ## Additional modes
   doInputRef: False
   inputRef: 'gnl|hbvnuc|AB064313_FT00000_C-G'
   doPerSamp: False
   
   ## RNAsim
   doSim: False
   sample_annotation_sm: RNAsim2/output/sampleAnnotation
   genomeId: gnl|hbvnuc|GQ924620_FT00000_C-C
   sampNum: "165000"
   pairedEndDist: "250"
   readLen: "150"
   mt: -mt
   mp: -mp 
   mtPos: "100"


RNAsim
######

RNAsim is a bare-bone RNA read simulator used prinicipally to generate reads for automatic integrations runs of the pipeline via github actions.
The user can run the program using reads generated by RNAsim by specifying the relevant parameters in the ``config/config.yaml``. For furher information on RNA see ``RNAsim2/readme.md``

SampleAnnotation 
################

The input reads are listed in a ``sampleAnnotation`` file. The corresponding parameter in the default config file points to a directory of simulated reads generated by RNAsim. you can change this to point to your input. The ``sampleAnnotation`` file requires a minimum of 4 columns per sample, see ``RNAsim2/output/sampleAnnotation``.


Run-modes
#########

The program can be run in three different modes: 1) infref: the reference genome is inferred from the aggregation of all samples, 2) inpt: samples are compared against a given genotype of HBV, and 3) per_Samp: the reference genome is inferred for each sample. By default all three modes are active. The inpt and per_samp modes can be excluded via the the paramters ``doInputRef`` sand ``doPerSamp`` in the config file.
 


Run snakemake
#############

Once the input files are specified and the run-mode is selected, run the pipeline by typing the following in a terminal at the project's root directory.

.. code-block:: console

   $ snakemake --cores 1 --use-conda

This will execute the pipeline localy. Please refer to snakemake documentation for job submission via SLURM or other schedulers.

Running snakemake with ``--forceall`` option
############################################

MultiQC adjusts filenames when it finds previous MultiQC output by numbering the reports. This causes snakemake to fail when it is run with --forceall option and previous MultiQC html report is present.



